### ISSUE TITLE
Nested Describe Functions Fail

### ISSUE THREAD

**Issue by FlippieCoetser** [meakbiyik/vscode-r-test-adapter#24]
According to `Testhat` Documentation, nesting `describe()` functions should be possible: [link](https://testthat.r-lib.org/reference/describe.html)

Below is a copy of the example in the documentation
```r
# Nested specs:
## code
addition <- function(a, b) a + b
division <- function(a, b) a / b

## specs
describe("math library", {
  describe("addition()", {
    it("can add two numbers", {
      expect_equal(1 + 1, addition(1, 1))
    })
  })
  describe("division()", {
    it("can divide two numbers", {
      expect_equal(10 / 2, division(10, 2))
    })
    it("can handle division by 0") #not yet implemented
  })
})
```
However, when trying to run the example in VS code, an error is thrown:

![image](https://user-images.githubusercontent.com/3672126/226442256-84164087-cbe7-4178-9a3d-990ff2d4a59c.png)

Removing the nesting like so:

```r
## specs
describe("math library", { 
  it("can add two numbers", {
    expect_equal(1 + 1, addition(1, 1))
  })
  it("can divide two numbers", {
    expect_equal(10 / 2, division(10, 2))
  })
  it("can handle division by 0") #not yet implemented
})
```
Works as expected.

![image](https://user-images.githubusercontent.com/3672126/226443246-889d6a8e-c8fa-4801-ab75-43e511d6c7f3.png)

**Comment by kubajal** [comment]
My proposal for an improvement of nested `describe`/`testthat`/`it` statements discovery in the `testthat` framework such that we don't have to implement ridiculous hacks in R Test Explorer (...or any other code editor...) was refused by the `testthat` maintainer in https://github.com/r-lib/testthat/issues/2118.
If you are a person needing this feature please upvote this comment https://github.com/r-lib/testthat/issues/2222#issuecomment-3199757490 and downvote

**Comment by kubajal** [comment]
tldr: upvote https://github.com/r-lib/testthat/issues/2184#issue-3299400053 and https://github.com/r-lib/testthat/pull/2220#issue-3325517319. Future maintainers in VSCode, RStudio and testthat will thank you.

The mechanism for running nested `describe`/`testthat`/`it` statments in VSCode has been partially enabled upstream and now awaits implementation on VSCode Test Explorer side. Let me provide my initial analysis of how to implement that and perhaps someone could implement that before I find spare time to do that.

The current approach in R Test Explorer for test discovery is that we have a custom [tree sitter parser for `testthat` tests](https://github.com/meakbiyik/vscode-r-test-adapter/blob/f81d934cc11344e7a834afd29b28c78ab5835237/src/testthat/parser.ts#L77). The problem is that tree sitter doesn't support arbitrary recursive grammar rules, see https://github.com/tree-sitter-grammars/tree-sitter-lua/issues/24#issue-1705577383. We need an improvement for discovery of `testthat` tests. This could be done in 2 ways.

#### 1. Easier dirtier way

Implement a workaround around tree-sitter grammar recursion limitation.
We could improve the [tree-sitter parser for `testthat` tests](https://github.com/meakbiyik/vscode-r-test-adapter/blob/f81d934cc11344e7a834afd29b28c78ab5835237/src/testthat/parser.ts#L77) such that there is a fixed maximal nesting level, let's say 5 or 10 as I don't believe anyone would write tests with higher nesting levels. The rest of the VSCode side logic stays essentially the same. This isn't a generic solution and this should be considered a workaround.

#### 2. Harder cleaner way

We could try to encourage the maintainers of `testthat` to properly implement test discovery on their side. We could then remove test discovery from our side which would otherwise require the workaround above.
By the way, there is a broader benefit to it because not only would it radically simplify the code on VSCode R Explorer side but also would enabled easier intergration of other IDEs as well. I believe that ironically RStudio would benefit from that too.
Unfortunately, the code changes impenenting better test discovery upstream in `testthat` was rejected in https://github.com/r-lib/testthat/pull/2220. Please support our case by upvoting https://github.com/r-lib/testthat/issues/2184#issue-3299400053 and https://github.com/r-lib/testthat/pull/2220#issue-3325517319.

**Comment by meakbiyik** [comment]
Thanks a lot for the summary @kubajal, totally agreed. If someone would like to tackle option 1, I would be happy to dust off my tree-sitter grammar knowledge, but it might be a tad challenging.
